arch/arm64/src/common/arm64_arch_timer.c

```diff
@@ -0,0 +1,254 @@
+{
+  uint64_t value;
+
+  value = read_sysreg(cntv_ctl_el0);
+
+  if (mask)
+    {
+      value |= CNTV_CTL_IMASK_BIT;
+    }
+  else
+    {
+      value &= ~CNTV_CTL_IMASK_BIT;
+    }
+
+  write_sysreg(value, cntv_ctl_el0);
+}
+
+static inline uint64_t arm64_arch_timer_count(void)
+{
+  return read_sysreg(cntvct_el0);
+}
+
+static inline uint32_t arm64_arch_timer_get_cntfrq(void)
+{
+  return read_sysreg(cntfrq_el0);
+}
+
+#ifdef CONFIG_SCHED_TICKLESS
+static int arm64_arch_timer_compare_isr(int irq, void *regs, void *arg)
+{
+  irqstate_t    flags;
+  uint64_t      curr_cycle;
+  uint32_t      delta_ticks;
+
+  UNUSED(regs);
+  UNUSED(arg);
+
+  flags = spin_lock_irqsave(&g_arch_timer_lock);
+
+  curr_cycle    = arm64_arch_timer_count();
+  delta_ticks   = (uint32_t)((curr_cycle - last_cycle) / cycle_per_tick);
+
+  last_cycle += delta_ticks * cycle_per_tick;
+
+  arm_arch_timer_set_irq_mask(true);
+
+  spin_unlock_irqrestore(&g_arch_timer_lock, flags);
+
+  nxsched_process_timer();
+  return OK;
+}
+
+#else
+
+static int arm64_arch_timer_compare_isr(int irq, void *regs, void *arg)
+{
+  uint64_t      curr_cycle;
+  uint32_t      delta_ticks;
+  uint64_t      next_cycle;
+
+  UNUSED(irq);
+  UNUSED(regs);
+  UNUSED(arg);
+
+  curr_cycle    = arm64_arch_timer_count();
+  delta_ticks   = (uint32_t)((curr_cycle - last_cycle) / cycle_per_tick);
+
+  last_cycle += delta_ticks * cycle_per_tick;
+
+  next_cycle = last_cycle + cycle_per_tick;
+
+  if ((uint64_t)(next_cycle - curr_cycle) < MIN_DELAY)
+    {
+      next_cycle += cycle_per_tick;
+    }
+
+  arm64_arch_timer_set_compare(next_cycle);
+  arm64_arch_timer_set_irq_mask(false);
+
+  nxsched_process_timer();
+  return OK;
+}
+
+#endif
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+#ifdef CONFIG_SMP
+/* Notes:
+ *
+ * The origin design for ARMv8-A timer is assigned private timer to
+ * every PE(CPU core), the ARM_ARCH_TIMER_IRQ is a PPI so it's
+ * should be enable at every core.
+ *
+ * But for NuttX, it's design only for primary core to handle timer
+ * interrupt and call nxsched_process_timer at timer tick mode.
+ * So we need only enable timer for primary core
+ *
+ * IMX6 use GPT which is a SPI rather than generic timer to handle
+ * timer interrupt
+ */
+
+void arm64_smp_timer_init(void)
+{
+  uint64_t curr_cycle;
+
+  /* set the initial status of timer0 of each secondary core */
+
+  curr_cycle = arm64_arch_timer_count();
+
+  arm64_arch_timer_set_compare(curr_cycle + cycle_per_tick);
+  arm64_arch_timer_enable(true);
+  up_enable_irq(ARM_ARCH_TIMER_IRQ);
+  arm64_arch_timer_set_irq_mask(false);
+}
+
+#endif
+
+uint64_t arm64_counter_read(void)
+{
+  return arm64_arch_timer_count();
+}
+
+void arm64_timer_delay(long mini_sec)
+{
+  unsigned int          timeout;
+  volatile unsigned int new_counter;
```

Comment:
why `volatile` is needed here?

Reply:
Else perhaps (current or future) compiler could optimize away `new_counter = arm64_counter_read();` in do..while()?

Reply:
Actually, I wonder if this may be subject to edge case if roll-over occurs.

Would this be better:

```
start_time = arm64_counter_read();
do
  {
    arch_nop();
  }
while ((arm64_counter_read() - start_time) < clock);
```

???

Reply:
Oh, arm64_timer_delay is a very early version to used at arm64_assert.c, only for debug propose
I think I perfer to use up_mdelay instead

Thanks for check